[Feature] #34 게시판 기본 CRUD 작성

csumin0825 opened the pull request:
close #34 

- Article - Board 매핑 ManyToMany로 변경

게시판 CRUD 작성
- 글 작성
  - 카테고리 별로 입력값이 달라서 PostMapping 따로 만들려다가 하나로 만들었습니다.
  - 글 작성 시 팀, 아티스트 해시태그로 입력하면 List<String>으로 변환 -> Board로 생성 예정
- 글 상세조회
- 글 수정
- 글 삭제

* 글 목록 조회는 Board, BoardCategory이랑 같이해야할 것 같아서 주석처리해놓았습니다


------------------------ 추가 수정 중

Board 관련 기능 삭제
- entity 삭제
- Article과 매핑 삭제
- BoardRepository 삭제

Hashtag Entity 생성
- Article과 매핑

게시글 CRUD 수정
  - builder 사용
  - controller 응답값 수정
  - 글 작성, 수정, 상세조회, 삭제 로직 수정
  - 글 목록조회 기능 구현
  - 글 작성, 수정에 해시태그 추가

** 다른 코드에서 Board 관련 코드 변경해야합니다!!!!!!!!!!!!!!!!!!!!
** Artist 등 다른 코드에서 Board 관련 코드오류가 있어서 현재 게시글 테스트를 못했습니다,,,,,,,

csumin0825 committed:
refactor: Article-Board ManyToMany 매핑으로 변경

csumin0825 committed:
feat: 글 작성(자유, 친목/동행), 글 상세 조회 구현

csumin0825 committed:
refactor: BoardCategory Entity 어노테이션 수정

csumin0825 committed:
feat: 게시판 CRUD 구현
  - 글 작성
  - 글 상세조회
  - 글 수정
  - 글 삭제

greedyBackEnd commented on src/main/java/com/happiday/Happi_Day/domain/controller/ArticleController.java:
```diff
@@ -0,0 +1,65 @@
+    private final ArticleService articleService;
+
+    // 글 작성
+    @PostMapping(value ="/{categoryId}/write", consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
```
RESTful API  설계 지침에 맞게 URL을 설계한다면, URL에 동작을 잘 표현하지는 않습니다. 
write 빼고 POST /api/v1/articles/{categoryId}/write → POST /api/v1/articles/{categoryId} 는 어떠실까욤?

greedyBackEnd commented on src/main/java/com/happiday/Happi_Day/domain/controller/ArticleController.java:
```diff
@@ -0,0 +1,65 @@
+    // 글 작성
+    @PostMapping(value ="/{categoryId}/write", consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
+    public void writeArticle(
+            @RequestPart(name="article") WriteArticleDto requestDto,
```
여러 @RequestPart를 받는 것 보다 DTO에 최대한의 정보를 넣고와 MultipartFile만 더해서 받는건 어떨가욤?

greedyBackEnd commented on src/main/java/com/happiday/Happi_Day/domain/controller/BoardController.java:
```diff
@@ -0,0 +1,23 @@
+
+    // 카테고리 생성
+    // 관리자만 접근 가능하도록
+    @PostMapping("/category/create")
```
앞서 설명드린 바와 같이 URL 설계에 create를 빼고 POST로 받는 건 어떨까요?

greedyBackEnd commented on src/main/java/com/happiday/Happi_Day/domain/entity/article/Article.java:
```diff
@@ -4,30 +4,29 @@
-    @ManyToOne(fetch = FetchType.LAZY)
-    @JoinColumn(name="board_id")
-    private Board board;
+    @ManyToMany
```
게시판과 게시글이 다대다가 되어야하는 것에는 어떤 이유가있을까욤?_?

greedyBackEnd commented on src/main/java/com/happiday/Happi_Day/domain/service/ArticleService.java:
```diff
@@ -0,0 +1,117 @@
+//    }
+
+    // 글 수정
+    public void updateArticle(Long articleId, WriteArticleDto dto, String artists, String teams, MultipartFile image, String address) throws IOException {
```
글이 수정되면 수정된 해당 게시글이 반환되는 건 어떨까욤?

greedyBackEnd commented on src/main/java/com/happiday/Happi_Day/domain/entity/board/BoardCategory.java:
```diff
@@ -3,19 +3,16 @@
 import java.util.List;
 
 @Entity
+@Data
```
엔티티에 @Data 어노테이션이 사용되는 것은 지양되고 있습니다.
대체품으로 빌더를 사용하거나, 상태 등의 특별한 경우 별도의 setter를 사용하시는 건 어떨까요?

greedyBackEnd reviewed (commented):
전체적으로 잘 짜신 것 같아요!
몇 가지 고려사항만 수정해주시면 좋을 것 같습니다.
추가적으로 jpa메서드 사용 시 .orElseThrow 등의 함수를 사용하면 좀 더 간결한 로직을 작성하실 수 있습니다! 수고하셨습니다~

csumin0825 committed:
feat: Board Entity 삭제

csumin0825 committed:
feat: Article-BoardCategory 매핑 수정

csumin0825 committed:
feat: Hashtag Entity 생성
  - Hashtag - Article 매핑

csumin0825 committed:
feat: BoardRepository 삭제

csumin0825 committed:
feat: 게시글 CRUD 수정
  - builder 사용
  - controller 응답값 수정
  - 글 작성, 수정, 상세조회, 삭제 로직 수정
  - 글 목록조회 기능 구현
  - 글 작성, 수정에 해시태그 추가

csumin0825 committed:
refactor: BoardCategory @Data 어노테이션 삭제

sy-k-98 commented on src/main/java/com/happiday/Happi_Day/domain/entity/article/Article.java:
```diff
@@ -40,15 +44,45 @@ public class Article extends BaseEntity {
     private List<Comment> comments = new ArrayList<>();
 
-    // 게시글 좋아요 매핑
+    // 게시글 좋아요 매핑정
```
매핑정...?  오타가 났네염

sy-k-98 commented on src/main/java/com/happiday/Happi_Day/domain/entity/article/Article.java:
```diff
@@ -40,15 +44,45 @@ public class Article extends BaseEntity {
     private List<User> scrapUsers = new ArrayList<>();
+
+    // 게시글 아티스트 매핑
+    @ManyToMany(mappedBy ="articleArtists")
```
별거는 아니지만 코드를 다 작성하고 컨트롤 알트 L을 누르면 자동 정렬 해줌니당

sy-k-98 reviewed (commented):
내용이 엄청 많아서 고생이 많으십니다.. 화이팅

yonzn0213 commented on src/main/java/com/happiday/Happi_Day/domain/service/ArticleService.java:
```diff
@@ -0,0 +1,133 @@
+    }
+
+    // 글 목록 조회
+    @Transactional
```
제가 알기론 클래스에 @Transactional(readOnly = true) 붙였을 때 조회하는 메서드들엔 @Transactional을 안 붙여도 되는 걸로 알고 있습니다! 작성, 수정, 삭제 같은 기능에만 붙이면 될거에요 아마,,,

2hapu reviewed (commented):
양이 많군요,, 제가 봤을 땐 다 된 것 같습니다 고생하셨어용!!!

csumin0825 committed:
refactor: 글 조회 기능에서 @Transactional 삭제

csumin0825 committed:
refactor: 오타 수정

csumin0825 committed:
refactor: 코드 정렬, BoardService 삭제

csumin0825 committed:
refactor: BoardController 삭제
